grpc-clients/kaas.go

```diff
@@ -0,0 +1,67 @@
+package clients
+
+import (
+	"fmt"
+
+	"github.com/RTradeLtd/config"
+	"github.com/RTradeLtd/grpc/dialer"
+	pb "github.com/RTradeLtd/grpc/krab"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// KaasClient is how we interface with the kaas grpc key manager
+type KaasClient struct {
+	pb.ServiceClient
+	conn *grpc.ClientConn
+}
+
+// NewKaasClient is used to instantiate our kaas client in primary or fallback mode
+func NewKaasClient(opts config.Services, fallback bool) (*KaasClient, error) {
+	var (
+		url      string
+		dialOpts []grpc.DialOption
+	)
+	if fallback {
+		if opts.Krab.Fallback.TLS.CertPath != "" {
+			creds, err := credentials.NewClientTLSFromFile(opts.Krab.Fallback.TLS.CertPath, "")
+			if err != nil {
+				return nil, fmt.Errorf("could not load tls cert: %s", err)
+			}
+			dialOpts = append(dialOpts,
+				grpc.WithTransportCredentials(creds),
+				grpc.WithPerRPCCredentials(dialer.NewCredentials(opts.Krab.AuthKey, true)))
+		} else {
+			dialOpts = append(dialOpts,
+				grpc.WithInsecure(),
+				grpc.WithPerRPCCredentials(dialer.NewCredentials(opts.Krab.AuthKey, false)))
+		}
+		url = opts.Krab.Fallback.URL
+	} else {
+		if opts.Krab.TLS.CertPath != "" {
```

Comment:
why duplicate all this code? it seems to me the only difference is the TLS config - maybe we could just assign the appropriate config based on `fallback`

Reply:
good point, how does this look
```golang
// NewKaasClient is used to instantiate our kaas client in primary or fallback mode
func NewKaasClient(opts config.Services, fallback bool) (*KaasClient, error) {
	var (
		dialOpts   []grpc.DialOption
		krabConfig config.Krab
	)
	if fallback {
		krabConfig.AuthKey = opts.Krab.Fallback.AuthKey
		krabConfig.KeystorePassword = opts.Krab.Fallback.KeystorePassword
		krabConfig.LogFile = opts.Krab.Fallback.LogFile
		krabConfig.TLS = opts.Krab.Fallback.TLS
		krabConfig.URL = opts.Krab.Fallback.URL
	} else {
		krabConfig = opts.Krab
	}
	if krabConfig.TLS.CertPath != "" {
		creds, err := credentials.NewClientTLSFromFile(krabConfig.TLS.CertPath, "")
		if err != nil {
			return nil, fmt.Errorf("could not load tls cert: %s", err)
		}
		dialOpts = append(dialOpts,
			grpc.WithTransportCredentials(creds),
			grpc.WithPerRPCCredentials(dialer.NewCredentials(krabConfig.AuthKey, true)))
	} else {
		dialOpts = append(dialOpts,
			grpc.WithInsecure(),
			grpc.WithPerRPCCredentials(dialer.NewCredentials(krabConfig.AuthKey, false)))
	}
	conn, err := grpc.Dial(krabConfig.URL, dialOpts...)
	if err != nil {
		return nil, err
	}
	return &KaasClient{
		conn:          conn,
		ServiceClient: pb.NewServiceClient(conn),
	}, nil
}
```

Reply:
I notice [here](https://github.com/RTradeLtd/config/blob/master/types.go#L175) that fallback is *exactly* the normal config - why not just make `Fallback` of type `Krab` and here we can have:

```go
if fallback {
    conf = opts.Krab.Fallback
} else {
    conf = opts.Krab
}
```

Reply:
even better!

Reply:
hmm that would lead to a recursive type

Reply:
probably dont put it together then

```
opts.Krab
opts.KrabFallback
```

Reply:
Excellent idea, heading to bed right now but will do this in the morning 👌

Reply:
Ok made the change, also realized one other spot where inconsistencies would show up during key export, so I configure the API to have clients for both kb1+kb2 and issue deletes to both instances